Make the shard day boundary timeZone configurable

hellium01 opened the pull request.

hellium01 commented:
This PR addresses the comment of @AlekseiS  on PR #9427
The timezone will be a setting on system level (removing table level support for now). Also added testing for shard splitting and  shard compaction especially during migration phase.  

Move temporal function to one place will make adding support for new timestamp format easier as well. Once everything is migrated to new timestamp(timestamp without timezone), we can deprecate the timezone setting easily.

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/StorageManagerConfig.java:
```diff
@@ -59,6 +61,7 @@
     private DataSize maxShardSize = new DataSize(256, MEGABYTE);
     private DataSize maxBufferSize = new DataSize(256, MEGABYTE);
     private int oneSplitPerBucketThreshold;
+    private String shardDayBoundaryTimeZone = TimeZone.getDefault().getID();
```
I think it should be explicitly set to UTC by default to match existing behavior.

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunctionFactory.java:
```diff
@@ -0,0 +1,132 @@
+        if (type == DATE) {
+            return new DateTemporalFunction();
+        }
+        else if (type.equals(TIMESTAMP)) {
```
Use comparisons consistently: you used '==' in the first comparison and 'equals' here. I think you need to use `equals` for types.

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunctionFactory.java:
```diff
@@ -0,0 +1,132 @@
+    public static class TimestampTemporalFunction
+            implements TemporalFunction
+    {
+        DateTimeZone timeZone;
```
private final

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunctionFactory.java:
```diff
@@ -0,0 +1,132 @@
+    public TemporalFunction createTemporalFunction(Type type)
+    {
+        if (type == DATE) {
+            return new DateTemporalFunction();
```
The classes you create are stateless, so you don't need to create a new one on every method call.

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorPageSink.java:
```diff
@@ -222,7 +229,7 @@ public void appendPage(Page page)
             for (int position = 0; position < page.getPositionCount(); position++) {
                 int bucket = bucketFunction.isPresent() ? bucketFunction.get().getBucket(bucketArgs, position) : 0;
-                int day = temporalFunction.isPresent() ? temporalFunction.get().getDay(temporalBlock, position) : 0;
+                int day = temporalFunction.isPresent() ? temporalFunction.get().getDay(session, temporalBlock, position) : 0;
```
Session was not necessary before, why is it needed now?

hellium01 replied:
This allows us to use getObject instead of assuming Timestamp format is UTC offset, so perhaps easier to expand later for new format?

electrum replied:
We don't need session here and should not be using `getObjectValue()` as that is extremely inefficient and is only used to return values to the client. We are not "assuming" the timestamp representation: it is well-defined for the type.

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunctionFactory.java:
```diff
@@ -0,0 +1,132 @@
+        @Override
+        public int getDay(Object min, Object max)
+        {
+            return ((Integer) min).intValue();
```
unnecessary unboxing

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunction.java:
```diff
@@ -0,0 +1,27 @@
+{
+    int getDay(ConnectorSession session, Block block, int position);
+
+    int getDay(Object min, Object max);
```
Why are parameters have type `Object` and not more specific type? Also, it's not clear from the method signature what this function is intended to do, so think of a better name for it and potentially a javadoc description.

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizerUtil.java:
```diff
@@ -158,13 +159,17 @@ private static ShardIndexInfo toShardIndexInfo(ShardMetadata shardMetadata, Opti
-
+        Optional<TemporalFunction> temporalFunction = Optional.ofNullable(Iterables.getFirst(shards, null))
+                .flatMap(ShardIndexInfo::getTemporalType)
+                .map(type -> temporalFunctionFactory.createTemporalFunction(type));
```
can be replaced with method reference

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunction.java:
```diff
@@ -0,0 +1,27 @@
+
+    int getDay(Object min, Object max);
+
+    Type getType();
```
If you chose the interface correctly then there should be no need for exposing this type. I think the client shouldn't depend on the type of temporal function being passed. Besides, it was working without this information before.

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunction.java:
```diff
@@ -0,0 +1,27 @@
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.type.Type;
+
+public interface TemporalFunction
```
Think of a better name

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunctionFactory.java:
```diff
@@ -0,0 +1,132 @@
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class TemporalFunctionFactory
```
In fact, in the simplest approach you don't need a factory. The logic is stateless, so you can pass the type as part of method call.

hellium01 replied:
Passing type as part of method call might mean we need to do type dispatching each time we call the getDay function, which might add unnecessary cost because we need to do this for each row of the page.

electrum replied:
I'm not concerned about an `if` check here, as that's insignificant compared to all the other work being done.

@AlekseiS, I'm interested in your idea. Can you explain more? Where would the time zone config go (which is now in the factory)?

AlekseiS replied:
The comment above was about making the class directly injectable. It should be possible given that it can be made essentially stateless (except for timezone config which is constant). In this case you don't need a factory and can directly use the "timezone shard splitter" class directly. The class will have only two simple methods and will be helpful to consolidate shard boundary timezone conversion logic.

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunctionFactory.java:
```diff
@@ -0,0 +1,132 @@
+            return TIMESTAMP;
+        }
+
+        private int determineDay(long rangeStart, long rangeEnd)
```
Ideally, we should keep this logic in `ShardOrganizerUtils` and only extract the part related to date boundary/timezone.

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorPageSink.java:
```diff
@@ -210,7 +217,7 @@ public PartitionedPageWriter()
 
             this.bucketFunction = bucketCount.isPresent() ? Optional.of(new RaptorBucketFunction(bucketCount.getAsInt(), bucketTypes)) : Optional.empty();
-            this.temporalFunction = temporalColumnType.map(TemporalFunction::create);
+            this.temporalFunction = temporalColumnType.map(type -> temporalFunctionFactory.getTemporalFunction(type));
```
I think this can use a method reference.

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorPageSink.java:
```diff
@@ -67,12 +69,15 @@
     private final PageWriter pageWriter;
 
     public RaptorPageSink(
+            ConnectorSession session,
```
Not needed, see below

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/StorageManagerConfig.java:
```diff
@@ -336,4 +340,17 @@ public StorageManagerConfig setOneSplitPerBucketThreshold(int oneSplitPerBucketT
+    }
+
+    @Config("storage.shard-day-boundary-time-zone")
+    @ConfigDescription("Day boundary to split the time zone")
```
"Time zone to use for computing day boundary for shards"

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardCompactionManager.java:
```diff
@@ -79,12 +79,13 @@
 
     @Inject
-    public ShardCompactionManager(@ForMetadata IDBI dbi, NodeManager nodeManager, ShardManager shardManager, ShardOrganizer organizer, StorageManagerConfig config)
+    public ShardCompactionManager(@ForMetadata IDBI dbi, NodeManager nodeManager, ShardManager shardManager, ShardOrganizer organizer, TemporalFunctionFactory temporalFunctionFactory, StorageManagerConfig config)
```
This list is long, split onto separate lines

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizationManager.java:
```diff
@@ -226,9 +231,9 @@ private boolean shouldRunOrganization(TableOrganizationInfo info)
     }
 
     @VisibleForTesting
-    static Set<OrganizationSet> createOrganizationSets(Table tableInfo, Collection<ShardIndexInfo> shards)
```
This method is only exposed for testing, so let's keep it static

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizerUtil.java:
```diff
@@ -116,14 +113,16 @@ private ShardOrganizerUtil() {}
+                            Optional<Type> temporalType = Optional.empty();
                             if (temporalColumn.isPresent()) {
                                 temporalRange = getShardRange(ImmutableList.of(temporalColumn.get()), resultSet);
+                                temporalType = temporalColumn.map(TableColumn::getDataType);
```
This should be moved below, so we don't have a `temporalType` without a `temporalRange`

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizerUtil.java:
```diff
@@ -158,13 +158,15 @@ private static ShardIndexInfo toShardIndexInfo(ShardMetadata shardMetadata, Opti
             return Multimaps.index(shards, shard -> shard.getBucketNumber().getAsInt()).asMap().values();
         }
-
+        Optional<TemporalFunction> temporalFunction = Optional.ofNullable(Iterables.getFirst(shards, null))
```
Add a check to the top of the method to exit early if the list is empty

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunction.java:
```diff
@@ -0,0 +1,24 @@
+{
+    int getDay(ConnectorSession session, Block block, int position);
+
+    int getDayFromRange(Tuple min, Tuple max);
```
Let's make this take `ShardRange`

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunctionFactory.java:
```diff
@@ -0,0 +1,127 @@
+
+    public static class DateTemporalFunction
+            implements TemporalFunction
+
```
Remove extra line

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunctionFactory.java:
```diff
@@ -0,0 +1,127 @@
+    @Inject
+    public TemporalFunctionFactory(StorageManagerConfig config)
+    {
+        this.timestampTemporalFunction = new TimestampTemporalFunction(requireNonNull(config.getShardDayBoundaryTimeZone(), "shardDayBoundaryTimeZone is null"));
```
Let's split this into two constructors, one that takes the config object and another that takes the time zone.

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunctionFactory.java:
```diff
@@ -0,0 +1,127 @@
+        if (type.equals(DATE)) {
+            return dateTemporalFunction;
+        }
+        else if (type.equals(TIMESTAMP)) {
```
`else` is redundant

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunctionFactory.java:
```diff
@@ -0,0 +1,127 @@
+        @Override
+        public int getDay(ConnectorSession session, Block block, int position)
+        {
+            return toIntExact(((SqlDate) DATE.getObjectValue(session, block, position)).getDays());
```
These methods should use the same code as before.

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/TestRaptorConnector.java:
```diff
@@ -167,12 +189,90 @@ public void testMaintenanceUnblockedOnStart()
     }
 
+    @Test
+    public void tesTemporalShardSplit()
```
Typo

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/TestRaptorConnector.java:
```diff
@@ -77,14 +99,14 @@ public void setup()
         ShardManager shardManager = createShardManager(dbi);
         StorageManager storageManager = createOrcStorageManager(dbi, dataDir);
-        StorageManagerConfig config = new StorageManagerConfig().setDataDirectory(dataDir);
+        StorageManagerConfig config = new StorageManagerConfig().setDataDirectory(dataDir).setShardDayBoundaryTimeZone("America/Los_Angeles");
```
Wrap the setters

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/TestRaptorConnector.java:
```diff
@@ -77,14 +99,14 @@ public void setup()
                 new RaptorSplitManager(connectorId, nodeSupplier, shardManager, false),
                 new RaptorPageSourceProvider(storageManager),
-                new RaptorPageSinkProvider(storageManager, new PagesIndexPageSorter(new PagesIndex.TestingFactory(false)), config),
+                new RaptorPageSinkProvider(storageManager, new PagesIndexPageSorter(new PagesIndex.TestingFactory(false)), new TemporalFunctionFactory(config), config),
```
This is getting long, let's wrap the args to `RaptorPageSinkProvider`

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/TestRaptorConnector.java:
```diff
@@ -167,12 +189,90 @@ public void testMaintenanceUnblockedOnStart()
+        assertSplitShard(TIMESTAMP, "2001-08-22 00:00:01.000", "2001-08-22 23:59:01.000", "America/Los_Angeles", 1);
+    }
+
+    private void assertSplitShard(Type temporalType, String min, String max, String userTimeZone, int numSplit)
```
I think we can do this test much simpler as a write query in `TestRaptorIntegrationSmokeTest` and a verification using the `$shard_uuid` column (see the bucket tests)

hellium01 replied:
Unfortunately, using IntegrationSmokeTest in this case is more difficult. First, changing timezone for different test is difficult because the session properties are shared between tests. Secondly, we validates result against H2 so under current implementation only UTC timezone will produce same result as H2.

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/storage/organization/TestCompactionSetCreator.java:
```diff
@@ -44,7 +47,7 @@
     private static final Table bucketedTemporalTableInfo = new Table(1L, OptionalLong.empty(), Optional.empty(), OptionalInt.of(3), OptionalLong.of(1), false);
 
-    private final CompactionSetCreator compactionSetCreator = new CompactionSetCreator(MAX_SHARD_SIZE, MAX_SHARD_ROWS);
+    private final CompactionSetCreator compactionSetCreator = new CompactionSetCreator(new TemporalFunctionFactory(new StorageManagerConfig().setShardDayBoundaryTimeZone("PST")), MAX_SHARD_SIZE, MAX_SHARD_ROWS);
```
Wrap the args to `CompactionSetCreator`

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/storage/organization/TestCompactionSetCreator.java:
```diff
@@ -103,22 +106,53 @@ public void testNonTemporalRowCountBasedOrganizationSet()
-        long day1 = Duration.ofDays(Duration.ofNanos(System.nanoTime()).toDays()).toMillis();
-        long day2 = Duration.ofDays(Duration.ofMillis(day1).toDays() + 1).toMillis();
-        long day3 = Duration.ofDays(Duration.ofMillis(day1).toDays() + 2).toMillis();
+        DateTimeZone localTimeZone = DateTimeZone.forTimeZone(TimeZone.getTimeZone("PST"));
```
Does this need to match the time zone used to create `CompactionSetCreator`? If so, make it a constant

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/storage/organization/TestCompactionSetCreator.java:
```diff
@@ -103,22 +106,53 @@ public void testNonTemporalRowCountBasedOrganizationSet()
+            throws Exception
+    {
+        DateTimeZone oldTimeZone = DateTimeZone.forTimeZone(TimeZone.getTimeZone("UTC"));
+        DateTimeZone newTimeZone = DateTimeZone.forTimeZone(TimeZone.getTimeZone("PST"));
```
Same comment about time zone constant

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/storage/organization/TestShardOrganizationManager.java:
```diff
@@ -192,11 +192,13 @@ private static ShardIndexInfo shardWithTemporalRange(int bucketNumber, ShardRang
     private ShardOrganizationManager createShardOrganizationManager(long intervalMillis)
     {
-        return new ShardOrganizationManager(dbi, "node1", createShardManager(dbi), createShardOrganizer(), true, new Duration(intervalMillis, MILLISECONDS));
+        StorageManagerConfig config = new StorageManagerConfig();
```
This is only used once, no need for a variable

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/storage/organization/TestShardOrganizationManager.java:
```diff
@@ -192,11 +192,13 @@ private static ShardIndexInfo shardWithTemporalRange(int bucketNumber, ShardRang
     {
-        return new ShardOrganizationManager(dbi, "node1", createShardManager(dbi), createShardOrganizer(), true, new Duration(intervalMillis, MILLISECONDS));
+        StorageManagerConfig config = new StorageManagerConfig();
+        return new ShardOrganizationManager(dbi, "node1", createShardManager(dbi), createShardOrganizer(), new TemporalFunctionFactory(config), true, new Duration(intervalMillis, MILLISECONDS));
```
Wrap arguments to `ShardOrganizationManager`

hellium01 commented:
Hi, @electrum @AlekseiS , thanks for reviewing. I updated the commit to address most of the comments. Also, in a separate commit, I removed the factory method and will do type dispatch under each call.

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/StorageManagerConfig.java:
```diff
@@ -336,4 +340,17 @@ public StorageManagerConfig setOneSplitPerBucketThreshold(int oneSplitPerBucketT
+
+    public DateTimeZone getShardDayBoundaryTimeZone()
+    {
+        return DateTimeZone.forTimeZone(TimeZone.getTimeZone(shardDayBoundaryTimeZone));
```
That's an interesting way of creating it. Is there a way to directly create DateTimeZone from a string?

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/RaptorTemporalFunction.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Override
+    public int getDayFromRange(ShardRange range)
```
I still suggest not pulling this logic into this class and instead only move a part which gets a day from a timestamp.

hellium01 replied:
As discussed offline, there are benefits to keep all time related logic to same place. Otherwise, if there will be future changes, we have to go back to ShardOrganizer to add support for new types ... etc.

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunction.java:
```diff
@@ -0,0 +1,24 @@
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.type.Type;
+
+public interface TemporalFunction
```
There's only one implementation and there is unlikely to be more going forward. You can consider removing the interface.

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/RaptorTemporalFunction.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Override
+    public int getDay(Type type, Block block, int position)
```
To clarify my comments on the previous reviews, you can still return  a function/interface if you want to avoid type checking on every call. Discussed offline.

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/RaptorTemporalFunction.java:
```diff
@@ -0,0 +1,102 @@
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class RaptorTemporalFunction
```
This can be named `TemporalFunction`

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizerUtil.java:
```diff
@@ -158,13 +160,12 @@ private static ShardIndexInfo toShardIndexInfo(ShardMetadata shardMetadata, Opti
         if (tableInfo.getBucketCount().isPresent() && !tableInfo.getTemporalColumnId().isPresent()) {
             return Multimaps.index(shards, shard -> shard.getBucketNumber().getAsInt()).asMap().values();
         }
-
```
Don't remove

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizerUtil.java:
```diff
@@ -144,11 +142,15 @@ private static ShardIndexInfo toShardIndexInfo(ShardMetadata shardMetadata, Opti
     {
+        if (shards.size() == 0) {
+            return ImmutableList.of();
+        }
```
Add a newline after this

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/storage/organization/TestCompactionSetCreator.java:
```diff
@@ -103,22 +111,52 @@ public void testNonTemporalRowCountBasedOrganizationSet()
-                shardWithTemporalRange(TIMESTAMP, day2, day2),
-                shardWithTemporalRange(TIMESTAMP, day1, day1),
-                shardWithTemporalRange(TIMESTAMP, day3, day3));
+                shardWithTemporalRange(TIMESTAMP, startOfDay1, startOfDay2 - 100),
```
What is `100`?

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/storage/organization/TestCompactionSetCreator.java:
```diff
@@ -103,22 +111,52 @@ public void testNonTemporalRowCountBasedOrganizationSet()
-        long day1 = Duration.ofDays(Duration.ofNanos(System.nanoTime()).toDays()).toMillis();
-        long day2 = Duration.ofDays(Duration.ofMillis(day1).toDays() + 1).toMillis();
-        long day3 = Duration.ofDays(Duration.ofMillis(day1).toDays() + 2).toMillis();
+        long today = Duration.ofNanos(System.nanoTime()).toDays();
```
This doesn't make sense. `System.nanoTime()` is effectively an arbitrary 64-bit number and can be positive or negative. Why would we turn it into a number of days?

electrum replied:
I'm confused why we are changing this test at all.

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/storage/organization/TestCompactionSetCreator.java:
```diff
@@ -103,22 +111,52 @@ public void testNonTemporalRowCountBasedOrganizationSet()
+    {
+        DateTimeZone oldTimeZone = UTC;
+        DateTimeZone newTimeZone = LOCAL_TIME_ZONE;
+        long today = Duration.ofNanos(System.nanoTime()).toDays();
```
Same comments as above

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/storage/organization/TestCompactionSetCreator.java:
```diff
@@ -103,22 +111,52 @@ public void testNonTemporalRowCountBasedOrganizationSet()
+    }
+
+    @Test
+    public void testTemporalCompactionTimeZoneMigration()
```
I'm inclined to say, let's leave this class alone entirely and add a new `TestTemporalFunction`. As long as the function is computing days correctly, this code will work.

hellium01 commented on presto-raptor/src/test/java/com/facebook/presto/raptor/TestRaptorConnector.java:
```diff
@@ -167,12 +194,91 @@ public void testMaintenanceUnblockedOnStart()
+                TimeZoneKey.getTimeZoneKey(userTimeZone),
+                ENGLISH,
+                System.currentTimeMillis(),
+                new RaptorSessionProperties(new StorageManagerConfig()).getSessionProperties(),
```
beginInsert below uses session property external_batch_id,  empty list will fail.

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunction.java:
```diff
@@ -0,0 +1,102 @@
+        }
+
+        if (type.equals(TIMESTAMP)) {
+            long millis = block.getLong(position, 0);
```
This is wrong. You should never read from blocks directly. Always use `Type` to access the block, as types own the data representation within a block.

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/storage/organization/TestTemporalFunction.java:
```diff
@@ -0,0 +1,149 @@
+    {
+        TemporalFunction temporalFunction = new TemporalFunction(PST);
+        BlockBuilder blockBuilder = DATE.createBlockBuilder(new BlockBuilderStatus(), 2);
+        DATE.writeLong(blockBuilder, 1);
```
These calls can be chained

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizerUtil.java:
```diff
@@ -147,8 +143,12 @@ private static ShardIndexInfo toShardIndexInfo(ShardMetadata shardMetadata, Opti
-    public static Collection<Collection<ShardIndexInfo>> getShardsByDaysBuckets(Table tableInfo, Collection<ShardIndexInfo> shards)
+    public static Collection<Collection<ShardIndexInfo>> getShardsByDaysBuckets(Table tableInfo, Collection<ShardIndexInfo> shards, TemporalFunction temporalFunction)
     {
+        if (shards.size() == 0) {
```
Use `isEmpty()`

electrum commented:
Commit message should say "time zone" not "timeZone"

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunction.java:
```diff
@@ -0,0 +1,102 @@
+        throw new IllegalArgumentException("Wrong type for shard range: " + type);
+    }
+
+    private int determineDay(long rangeStart, long rangeEnd)
```
This method should be static

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunction.java:
```diff
@@ -0,0 +1,102 @@
+        }
+
+        if (type.equals(TIMESTAMP)) {
+            Long minValue = timeZone.convertUTCToLocal((Long) getOnlyElement(min.getValues()));
```
Don't box here

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/storage/organization/TestShardOrganizationManager.java:
```diff
@@ -60,6 +61,7 @@
     private static final Table temporalTableInfo = new Table(1L, OptionalLong.empty(), Optional.empty(), OptionalInt.empty(), OptionalLong.of(1), true);
 
     private static final List<Type> types = ImmutableList.of(BIGINT, VARCHAR, DATE, TIMESTAMP);
+    private static final TemporalFunction TEMPORAL_FUNCTION = new TemporalFunction(new StorageManagerConfig());
```
This should use the simple constructor

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/storage/organization/TestShardOrganizationManager.java:
```diff
@@ -197,6 +198,12 @@ private static ShardIndexInfo shardWithTemporalRange(int bucketNumber, ShardRang
+                "node1",
+                createShardManager(dbi),
+                createShardOrganizer(),
+                new TemporalFunction(new StorageManagerConfig()),
```
This can use the constant

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/storage/organization/TestShardOrganizationManager.java:
```diff
@@ -183,7 +184,7 @@ private static ShardIndexInfo shardWithSortRange(int bucketNumber, ShardRange so
     }
 
-    private static ShardIndexInfo shardWithTemporalRange(int bucketNumber, ShardRange sortRange, ShardRange temporalRange)
+    private static ShardIndexInfo shardWithTemporalRange(Type temporalType, int bucketNumber, ShardRange sortRange, ShardRange temporalRange)
```
This parameter is unused

electrum commented:
Merged, thanks!
